### PR TITLE
perf: reduce APK size with ProGuard enabled

### PR DESCRIPTION
**Pull request by hurali97**
### Details

This PR enables minification and resource shrinking for android using ProGuard. This gives us **10.4 MB** reduction in Download Size.

Before:

  | APK Size MB | Download Size MB
-- | -- | --
All architectures | 127 | 123.3
  |   |  

After:

  | APK Size MB | Download Size MB
-- | -- | --
All architectures | 116.5 | 112.9
  |   |  

### Fixed Issues

$ https://github.com/Expensify/App/issues/34222
PROPOSAL: https://github.com/Expensify/App/issues/33070#issuecomment-1875099319

### Tests

- Install the APK
- Launch the app and use normally
- The app shouldn't crash and shouldn't miss any functionality

- [X] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/47336142/8b320a4e-8159-487e-9555-8061187625de

</details>

<details>
<summary>Android: mWeb Chrome</summary>

N/A

</details>

<details>
<summary>iOS: Native</summary>

N/A

</details>

<details>
<summary>iOS: mWeb Safari</summary>

N/A

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

N/A

</details>

<details>
<summary>MacOS: Desktop</summary>

N/A

</details>

### CODE REVIEW

**Comment by hurali97** [comment]
It would be ideal to NOT MERGE this PR until the QA completes their testing and confirms that no crashes or loss of functionality is noticed. 

We can move forward with creating the builds for this PR and kick off the QA testing.

cc: @roryabraham @mountiny

**Comment by mountiny** [comment]
Running a build here https://github.com/Expensify/App/actions/runs/7474984009

**Review by Julesssss — DISMISSED** [review]
Testing well for me, though we'll need to keep a close eye out for regressions once this is merged as it removes and obfuscates code.

**Comment by Julesssss** [comment]
One super minor thing I noticed is that the AdHoc badge doesn't display, but I haven't checked previous AdHoc builds in a while...

Screenshot_20240110_132855

**Comment by mountiny** [comment]
@shubham1206agra will do some initial testing and we would probably ask QA to do a full regression on the android build before merging too

**Comment by mountiny** [comment]
@shubham1206agra will do some initial testing and we would probably ask QA to do a full regression on the android build before merging too

**Comment by shubham1206agra** [comment]
image

Please check notification here

**Comment by mountiny** [comment]
Is it the logo? that might be case for other adhoc builds too, that would be good to check

**Comment by shubham1206agra** [comment]
I think the App is working fine. Except the logo, crossed check with other adhoc build where it works fine.
Note - Please check the permissions area as I got a phantom permission while doing distance request in confirmation page.

cc @mountiny @Julesssss

**Comment by shubham1206agra** [comment]
https://github.com/Expensify/App/assets/58412969/d654d136-c0dd-4165-82e6-28971b2211ed

**Comment by shubham1206agra** [comment]
@hurali97 Please address this https://github.com/Expensify/App/pull/34218#issuecomment-1884869946

**Comment by mountiny** [comment]
@shubham1206agra did you confirm that does not happen on other Android adhoc builds?

**Comment by shubham1206agra** [comment]
Yes

**Comment by hurali97** [comment]
> @hurali97 Please address this [#34218 (comment)](https://github.com/Expensify/App/pull/34218#issuecomment-1884869946)

@shubham1206agra It's fixed now.

Screenshot 2024-01-11 at 5 16 14 PM

**Comment by shubham1206agra** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/58412969/d654d136-c0dd-4165-82e6-28971b2211ed

</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by shubham1206agra** [comment]
@hurali97 Can you confirm the final size of apk now?

**Comment by hurali97** [comment]
> @hurali97 Can you confirm the final size of apk now?

Yeah sure.


After adding keep rules for resources: 

  | APK Size MB | Download Size MB
-- | -- | --
All architectures | 116.5 | 113
  |   |  

There's 0.1 MB increase in Download Size and APK Size remains intact. We still save 10.3 MB in Download size.

File: android/app/src/main/res/raw/keep.xml
Comment (Julesssss) [inline android/app/src/main/res/raw/keep.xml:5]: This is a small selection of resources, did we get this list from anywhere specific?

Comment (hurali97) [reply android/app/src/main/res/raw/keep.xml:5]: I inspected the res directory and all the resources that were used in the codebase by search, were added in keep rule. For example, `drawable/datepicker_background.xml` wasn't being referenced anywhere, so I didn't add it to the keep rule. 

So there's no prescription that says what to keep but anything that's referenced and is being used can be kept. I also tested on build variants like production and adhoc and I get the correct assets as our keep xml file is inherited for each build variant.

https://developer.android.com/build/shrink-code#keep-resources

Screenshot 2024-01-11 at 6 08 06 PM

Comment (hurali97) [reply android/app/src/main/res/raw/keep.xml:5]: Resources in `react-native/assets` aren't touched here, so that's why they aren't added in the keep rules.

**Review by shubham1206agra — APPROVED** [review]
LGTM!

File: android/app/src/main/res/raw/keep.xml
Comment (Julesssss) [reply android/app/src/main/res/raw/keep.xml:5]: > I also tested on build variants like production and adhoc and I get the correct assets as our keep xml file is inherited for each build variant.

Okay cool, that's what I was worried about as it's not clear that alternative resourses would be kept. For example: `drawable-xxxhdpi/bootsplash_logo.png`

I'm just generating another test build to confirm that we're not defaulting to the lower res drawables.

**Comment by mountiny** [comment]
Making a new test build

**Review by Julesssss — APPROVED** [review]
LGTM, was unable to find any broken features

**Comment by Julesssss** [comment]
Oh wait, did we confirm the logo issue doesn't exist on other AdHoc builds?

> I think the App is working fine. Except the logo, crossed check with other adhoc build where it works fine.

**Comment by Julesssss** [comment]
One other thing, did this PR introduce the regression where Android build information isn't shown for AdHoc downloads?

Screenshot 2024-01-19 at 14 16 10

**Comment by shubham1206agra** [comment]
> Oh wait, did we confirm the logo issue doesn't exist on other AdHoc builds?
> 
> > I think the App is working fine. Except the logo, crossed check with other adhoc build where it works fine.

I have checked this. Works fine for me.

**Comment by Julesssss** [comment]
I'm still seeing it on a Pixel 6a, perhaps this depends on DP?

Screenshot_20240119-144503

**Comment by mountiny** [comment]
@Julesssss I have that on my iPhone adhoc build, other one too, not related

**Comment by shubham1206agra** [comment]
> I'm still seeing it on a Pixel 6a, perhaps this depends on DP?
> 
> Screenshot_20240119-144503

This is a known problem in other builds too

**Comment by mountiny** [comment]
I think this is good to go then!

**Review by mountiny — APPROVED** [review]
Ah why is this still not merged... going to merge it

**Pull request merged**
